app/routes.py

```diff
@@ -1,2 +1,239 @@
+                tasks_response.append({
+                    "id": task.task_id,
+                    "goal_id": task.goal_id,
+                    "title": task.title,
+                    "description": task.description,
+                    "is_complete": bool(task.completed_at)
+                })
+            response_body = {
+                "id": goal.goal_id,
+                "title": goal.title,
+                "tasks" : tasks_response
+            }
+            return jsonify(response_body), 200
+
+
+@tasks_bp.route("",methods=["GET","POST"])
+def handle_task():
+    if request.method == "GET": 
+
+        sorting_task = request.args.get('sort') 
+        list = None
+        if sorting_task == "desc":
+            list = Task.query.order_by(Task.title.desc()) # descending method
+        elif sorting_task == "asc":
+            list = Task.query.order_by(Task.title.asc()) # ascending method
+        else: 
+            list = Task.query.all()
+        tasks_response = []
+        for task in list:
+            tasks_response.append({
+            "id": task.task_id,
+            "title": task.title,
+            "description": task.description,
+            "is_complete" : False
+            if task.completed_at == None else True
+            })
+
+        return jsonify(tasks_response)
+
+    elif request.method == "POST":
+        request_body = request.get_json()
+        if "title" not in request_body or "description" not in request_body or "completed_at" not in request_body:
+            return{
+                "details": "Invalid data"
+            },400
+
+        new_task = Task (
+            title=request_body["title"],
+            description=request_body["description"],
+            completed_at=request_body["completed_at"]
+        )
+
+        db.session.add(new_task)
+        db.session.commit()
+
+        return {
+            "task":new_task.task_dict()
+        }, 201
+
+def update_completion(task_id, value):
+    task = Task.query.get(task_id)
+    
+    if not task:
+        return("Task not found",404)
+    
+    task.completed_at = value 
+
+    return {
+        "task":task.task_dict()
+    }, 200 
+
+@tasks_bp.route("/<task_id>/mark_complete", methods=["PATCH"])
+def mark_complete(task_id):
+    return update_completion(task_id, datetime.now())
+
+@tasks_bp.route("/<task_id>/mark_incomplete", methods =["PATCH"])
+def mark_incomplete(task_id):
+    return update_completion(task_id,None)
+
+
+@tasks_bp.route("/<task_id>", methods=["GET","PUT","DELETE"])
+def handle_task_get(task_id):
+    task = Task.query.get(task_id)
+    if task == None:
+        return ("", 404)
+
+    if request.method == "GET":
+        response_body = {}
+        response_body["task"] = task.task_dict()
+
+        return jsonify(response_body)
+
+    if request.method == "PUT":
+        form_data = request.get_json()
+
+        task.title = form_data["title"]
+        task.description = form_data["description"]
```

Comment:
We should be sure that the same fields required for POSTing are included here for PUT. PUT _replaces_ the value for the supplied task id, so we should ensure that all of the values required to represent a `Task` are provided with the same restrictions as we had when creating it in the first place.

At the least, we should ensure that the keys are present in the dictionary before trying to access them, otherwise this request would cause a `KeyError`.